homeassistant/config_entries.py

```diff
@@ -1051,6 +1066,64 @@ def _async_discovery(self) -> None:
         )
 
 
+class ConfigEntryItems(UserDict[str, ConfigEntry]):
+    """Container for config items, maps config_entry_id -> entry.
+
+    Maintains two additional indexes:
+    - domain -> list[ConfigEntry]
+    - domain -> unique_id -> ConfigEntry
+    """
+
+    def __init__(self) -> None:
+        """Initialize the container."""
+        super().__init__()
+        self._domain_index: dict[str, list[ConfigEntry]] = {}
+        self._domain_unique_id_index: dict[str, dict[str, ConfigEntry]] = {}
+
+    def values(self) -> ValuesView[ConfigEntry]:
+        """Return the underlying values to avoid __iter__ overhead."""
+        return self.data.values()
+
+    def __setitem__(self, key: str, entry: ConfigEntry) -> None:
+        """Add an item."""
+        data = self.data
+        if key in data:
+            self._unindex_entry(key)
```

Comment:
Unconditionally call unindex entry and move the check for existense into function `if not entry := self.data.get()`. Avoids the double key lookup.

Reply:
I think it might make sense to raise ValueError here since we should never overwrite one

Reply:
added a guard to prevent it from being added twice 162567d52d6

Reply:
There was an outer guard in `async_add`, but better to make sure nothing can corrupt the state of the internals